include/kernels/reduce_core.cuh

```diff
@@ -3,60 +3,78 @@
 #include <color_spinor_field_order.h>
 #include <blas_helper.cuh>
 #include <reduce_helper.h>
-#include <fast_intdiv.h>
+#include <reduction_kernel.h>
 
 namespace quda
 {
 
   namespace blas
   {
 
-    template <typename store_t, int N, typename y_store_t, int Ny, typename Reducer_>
+    /**
+       @brief Parameter struct for generic reduction blas kernel.
+       @tparam real_ The precision of the calculation
+       @tparam n_ The number of real elements per thread
+       @tparam store_t Default store type for the fields
+       @tparam N Default field vector i/o length
+       @tparam y_store_t Store type for the y fields
+       @tparam Ny Y-field vector i/o length
+       @tparam Reducer_ Functor used to operate on data
+    */
+    template <typename real_, int n_, typename store_t, int N, typename y_store_t, int Ny, typename Reducer_>
     struct ReductionArg : public ReduceArg<typename Reducer_::reduce_t> {
+      using real = real_;
+      static constexpr int n = n_;
       using Reducer = Reducer_;
+      using reduce_t = typename Reducer_::reduce_t;
       Spinor<store_t, N> X;
       Spinor<y_store_t, Ny> Y;
       Spinor<store_t, N> Z;
       Spinor<store_t, N> W;
       Spinor<store_t, N> V;
       Reducer r;
 
-      const int length;
+      const int length_cb;
       const int nParity;
-      const int_fastdiv gridSize;
 
       ReductionArg(ColorSpinorField &x, ColorSpinorField &y, ColorSpinorField &z, ColorSpinorField &w,
-                   ColorSpinorField &v, Reducer r, int length, int nParity, TuneParam &tp) :
-        ReduceArg<typename Reducer_::reduce_t>(),
+                   ColorSpinorField &v, Reducer r, int length, int nParity) :
         X(x),
         Y(y),
         Z(z),
         W(w),
         V(v),
         r(r),
-        length(length),
-        nParity(nParity),
-        gridSize(tp.grid.x * tp.block.x / nParity)
-      { ; }
+        length_cb(length / nParity),
+        nParity(nParity)
+      {
+        this->threads = dim3(length, 1, 1);
```

Comment:
Per a conversation with @maddyscientist on Slack, this `dim3` can be made a constructor argument to the base class `ReduceArg` (defined in target sub-directories), etc, propagating down to the base `kernel_param` class which accepts a `dim3` in the constructor.

Reply:
I've applied this change in 1831c77b1, and all reducer kernels now follow this.